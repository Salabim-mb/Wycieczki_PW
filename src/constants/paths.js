export const paths = {
	DASHBOARD: {
		name: "Strona główna",
		path: "/",
	},
	BLOG: {
		name: 'Blog',
		path: '/:blog',
	},
	BLOG_CATEGORIES: {
		name: "Kategorie",
		path: "/blog/:category",
		redirect: (newCategory) => `/blog/${newCategory}`,
	},
	BLOG_ENTRY: {
		name: "Wpis",
		path: "/blog/:category/:entry_id",
		redirect: (category, entryId) => `/blog/${category}/${entryId}`,
	},
	POST_CREATOR: {
		name: 'Kreator wpisów',
		path: '/blog/post-creator/:id?',
		redirect: (entryId) => `/blog/post-creator/${entryId}`
	},
	CREATOR_CATEGORY: {
		name: "Krator kategorii",
		path: "/blog/category-creator/:id?",
		redirect: (categoryId) => `/blog/category-creator/${categoryId}`
	}
}