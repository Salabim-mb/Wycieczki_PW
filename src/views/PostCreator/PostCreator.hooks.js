import { useMutation, useQuery } from 'react-query';
import {
	sendPost,
	sendAttachments,
	getTopics,
	deleteAttachments,
	deleteAttachment,
} from 'views/PostCreator/PostCreator.api';

export const usePostMutation = (token, prevId) =>
	useMutation(postData => sendPost(token, postData, prevId));
export const useAttachmentsMutation = token =>
	useMutation(attachmentList => sendAttachments(token, attachmentList));
export const useTopicsQuery = () => useQuery('topics', getTopics);
export const useFileDeletion = token => useMutation(data => deleteAttachments(token, data));
export const useImageDeletion = token => useMutation(data => deleteAttachment(token, data, true));
