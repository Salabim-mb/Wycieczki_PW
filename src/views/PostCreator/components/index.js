export { default as PostForm } from './PostForm';
export { default as ImageAdapter } from './ImageAdapter';
export { default as TopicSelector } from './TopicSelector';
export { default as AttachmentsList } from './AttachmentsList';
export { default as PostTilePreview } from './PostTilePreview';
export { default as HeaderPreview } from './HeaderPreview';
export { default as HeaderInput } from './HeaderInput';
export { default as PostEditor } from './PostEditor';
