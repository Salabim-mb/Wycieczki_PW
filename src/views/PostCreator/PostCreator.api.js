import paths from 'constants/api';

export const getPost = async (token, id) => {
	if (id) {
		// fetch post
		const url = `${paths.BLOG_POST}${id}/`;

		const headers = {};

		const response = await fetch(url, { headers, method: 'GET' });

		if (response.status === 200) {
			return response.json();
		}
		throw new Error('');
	}
	// return empty if id is not provided
	return {};
};

export const uploadPhoto = async (token, id, file, signal, flag, attachment) => {
	const url = `${attachment ? paths.BLOG_DOWNLOADABLE : paths.BLOG_IMAGE}`;

	const formData = new FormData();
	formData.append('reservation', id);
	formData.append(attachment ? 'file' : 'image', file, file.name);

	if (flag) {
		formData.append(flag, true);
	}

	const headers = {
		Authorization: `${token}`,
	};

	const response = await fetch(url, { headers, method: 'POST', signal, body: formData });

	if (response.status === 201) {
		return response.json();
	}
	throw new Error('');
};

export const sendPost = async (token, post, id) => {
	const url = `${paths.BLOG_POST}${id ? `${id}/` : ''}`;

	const headers = {
		Authorization: `${token}`,
		'Content-Type': 'application/json',
	};

	const postNew = { ...post };

	if (post.cover && typeof post.cover !== 'string') {
		const cover = await uploadPhoto('', post.reservation, postNew.cover, null, 'is_cover', false);
		postNew.cover = `${paths.PLAIN}${cover.image_url}`;
	}

	if (post.header && typeof post.header !== 'string') {
		const header = await uploadPhoto(
			'',
			post.reservation,
			postNew.header,
			null,
			'is_header',
			false,
		);
		postNew.header = `${paths.PLAIN}${header.image_url}`;
	}

	const method = id ? 'PATCH' : 'POST';
	const expectedResponseStatus = id ? 200 : 201;

	const response = await fetch(url, { headers, method, body: JSON.stringify(postNew) });

	if (response.status !== expectedResponseStatus) {
		throw new Error('');
	}
};

export const uploadAttachment = async (token, id, file) =>
	uploadPhoto(token, id, file, null, null, true);

export const sendAttachments = async (token, attachmentsObject) => {
	const { attachments, reservation } = attachmentsObject;
	await Promise.all(
		attachments?.map(({ file }) => file && uploadAttachment(token, reservation, file)),
	);
};

export const deleteAttachment = async (token, id, isImage) => {
	const startOfUrl = isImage ? paths.BLOG_IMAGE : paths.BLOG_DOWNLOADABLE;
	const url = `${startOfUrl}${id}/`;

	const headers = {
		Authorization: `${token}`,
	};

	const response = await fetch(url, { headers, method: 'DELETE' });

	if (response.status !== 204) {
		throw new Error('');
	}
	return response;
};

export const deleteAttachments = async (token, files) => {
	const imagesToDelete = files?.images.filter(
		({ image_url: imageUrl }) => !files.content.includes(imageUrl),
	);

	let attachmentsToDelete = [];

	if (files.prevAttachments) {
		attachmentsToDelete = files.prevAttachments.filter(
			e => !files.attachments.some(item => item.id === e.id),
		);
	}

	await Promise.all(imagesToDelete.map(({ id }) => deleteAttachment(token, id, true)));

	await Promise.all(attachmentsToDelete.map(({ id }) => deleteAttachment(token, id, false)));
};

export const reserveSpace = async token => {
	const url = `${paths.BLOG_RESERVATION}`;

	const headers = {
		Authorization: `${token}`,
	};

	const response = await fetch(url, { headers, method: 'POST' });

	if (response.status === 201) {
		return response.json();
	}
	throw new Error('');
};

export const getReservation = async id => {
	if (id) {
		const url = `${paths.BLOG_RESERVATION}${id}/`;

		const headers = {};

		const response = await fetch(url, { headers, method: 'GET' });

		if (response.status === 200) {
			return response.json();
		}
		throw new Error('');
	}
	// return empty if id is not provided
	return {};
};

export const getTopics = async () => {
	const url = `${paths.BLOG_TOPICS}`;

	const headers = {};

	const response = await fetch(url, { headers, method: 'GET' });

	if (response.status !== 200) {
		throw new Error('błąd');
	}

	return response.json();
};
