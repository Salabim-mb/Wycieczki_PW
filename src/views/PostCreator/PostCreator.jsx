import React from 'react';
import { useParams } from 'react-router-dom';
import { AlertInfo } from 'components';
import { Container, Card, CardHeader, CardContent } from '@material-ui/core';

import { PostForm } from './components';
import { usePostQuery } from './PostCreator.hooks';

const PostCreator = () => {
	const { id } = useParams();

	const { isLoading, isError, error, data } = usePostQuery(id);

	return (
		<Container>
			<Card>
				<CardHeader title="StwÃ³rz wpis" />
				<CardContent>
					<PostForm post={data && Object.keys(data).length === 0 ? null : data} id={id} />
					<AlertInfo isLoading={isLoading} isError={isError}>
						{error?.message}
					</AlertInfo>
				</CardContent>
			</Card>
		</Container>
	);
};

export default PostCreator;
