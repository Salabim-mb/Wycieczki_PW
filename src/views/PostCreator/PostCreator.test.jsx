import React from 'react';
import { render } from '@testing-library/react';
import { MemoryRouter } from 'react-router-dom';
import paths from 'constants/api';
import PostCreator from './PostCreator';
import { usePostQuery } from './PostCreator.hooks';

jest.mock('./components', () => ({
	PostForm: () => <div>formularz ca≈Çy</div>,
}));

jest.mock('views/PostCreator/PostCreator.hooks', () => ({
	usePostQuery: jest.fn(),
}));

jest.mock('react-router-dom', () => ({
	...jest.requireActual('react-router-dom'),
	useParams: () => ({
		id: 'abc',
	}),
}));

describe('PostCreator', () => {
	beforeEach(() => {
		jest.clearAllMocks();

		usePostQuery.mockImplementation(() => ({
			isSuccess: true,
			data: {
				content: 'aaa',
				cover: paths.PLAIN,
				downloadables: [],
				header: '',
				images: [],
				reservation: 100,
				summary: 'aaa',
				title: 'Sport',
			},
			isError: false,
			isLoading: false,
		}));
	});

	it('should match snapshot', async () => {
		const { container } = render(
			<MemoryRouter>
				<PostCreator />
			</MemoryRouter>,
		);

		expect(container).toMatchSnapshot();
	});
});
