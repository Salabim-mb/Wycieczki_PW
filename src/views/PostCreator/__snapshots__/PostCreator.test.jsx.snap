// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostCreator should match snapshot 1`] = `
<div>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <div
      class="MuiPaper-root MuiCard-root MuiPaper-elevation1 MuiPaper-rounded"
    >
      <div
        class="MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiCardHeader-title MuiTypography-h5 MuiTypography-displayBlock"
          >
            Stwórz wpis
          </span>
        </div>
      </div>
      <div
        class="MuiCardContent-root"
      >
        <div>
          formularz cały
        </div>
      </div>
    </div>
  </div>
</div>
`;
